### PR TITLE
[REVIEW] Remove usage of string based `set_dtypes` for `csv` & `json` readers

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR removes usage of `string` based `set_dtypes` in both `CSV` & `JSON` readers.

Partially addresses: https://github.com/rapidsai/cudf/issues/8240

### CODE REVIEW

File: python/cudf/cudf/_lib/csv.pyx [hunk 1]
```diff
@@ -232,37 +240,67 @@ cdef csv_reader_options make_csv_reader_options(
...
+                if col_dtype in {"hex", "hex64"}:
+                    c_hex_col_indexes.push_back(index)
+                    c_dtypes_list.push_back(
+                        _get_cudf_data_type_from_dtype('int64')
+                    )
+                elif col_dtype == "hex32":
+                    c_hex_col_indexes.push_back(index)
+                    c_dtypes_list.push_back(
+                        _get_cudf_data_type_from_dtype('int32')
+                    )
```
Comment (vuule) [inline python/cudf/cudf/_lib/csv.pyx:292]: variations of this block appear three times, wondering if the core logic can be factored out (namely, the hex->int mapping)

File: python/cudf/cudf/_lib/csv.pyx [hunk 2]
```diff
@@ -494,29 +532,38 @@ def _get_cudf_compatible_str_from_dtype(dtype):
...
-            "timestamp[us]", "timestamp[s]", "timestamp[ms]", "timestamp[ns]",
-            "date64"
-        }
-    ):
-        return str(dtype)
-    pd_dtype = pd.core.dtypes.common.pandas_dtype(dtype)
-
-    if pd_dtype in cudf.utils.dtypes.pandas_dtypes_to_cudf_dtypes:
-        return str(cudf.utils.dtypes.pandas_dtypes_to_cudf_dtypes[pd_dtype])
-    elif isinstance(pd_dtype, np.dtype) and pd_dtype.kind in ("O", "U"):
```
Comment (vuule) [inline python/cudf/cudf/_lib/csv.pyx:510]: is this still supported (left side of the diff)?

File: python/cudf/cudf/tests/test_json.py [hunk 3]
```diff
@@ -278,7 +278,7 @@ def test_json_lines_byte_range(json_input):
 )
 def test_json_lines_dtypes(json_input, dtype):
     df = cudf.read_json(json_input, lines=True, dtype=dtype)
-    assert all(df.dtypes == ["float32", "int32", "int16"])
+    assert all(df.dtypes == ["float64", "int64", "int16"])
```
Comment (vuule) [inline python/cudf/cudf/tests/test_json.py:281]: why this change?

**Review by vuule — COMMENTED** [review]
Minor suggestions/questions.

File: python/cudf/cudf/_lib/csv.pyx (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/_lib/csv.pyx:510]: Yes, that is being taken care of in `cudf.dtype` call.

File: python/cudf/cudf/tests/test_json.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_json.py:281]: So until now if the user passed `"float"` or `"int"`, csv reader would map them to `"float32"` or `"int32"` respectively. But that is not the general convention with `cudf`/`numpy`/`pandas`, `int` translates to `int64` and `float` translates to `float64`. 

```python
>>> import numpy as np
>>> np.dtype('float')
dtype('float64')
>>> np.dtype('int')
dtype('int64')
```

File: python/cudf/cudf/_lib/csv.pyx (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/_lib/csv.pyx]: Moved the type mapping to a separate dict that helped to remove lot of the code.

Comment (galipremsagar) [reply python/cudf/cudf/_lib/csv.pyx]: Moved the type mapping to a separate dict that helped to remove lot of the code.

File: python/cudf/cudf/tests/test_json.py (hunk 3 above)
Comment (vuule) [reply python/cudf/cudf/tests/test_json.py:281]: now I'm even more glad to remove the string-based dtype API from libcudf :) This would be either inconsistent or counterintuitive there.

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_json.py:281]: Yup, removing string based dtype selection surfaced this issue :tada:

File: python/cudf/cudf/_lib/csv.pyx (hunk 1 above)
Comment (vuule) [reply python/cudf/cudf/_lib/csv.pyx]: That looks great!

File: python/cudf/cudf/_lib/csv.pyx [hunk 4]
```diff
@@ -232,37 +246,50 @@ cdef csv_reader_options make_csv_reader_options(
...
             for k, v in dtype.items():
-                c_dtypes.push_back(
-                    str(
-                        str(k)+":"+
-                        _get_cudf_compatible_str_from_dtype(v)
-                    ).encode()
-                )
+                col_type = v
+                if v in CSV_HEX_TYPE_MAP:
+                    col_type = CSV_HEX_TYPE_MAP.get(v, v)
```
Comment (charlesbluca) [inline python/cudf/cudf/_lib/csv.pyx:252]:
```suggestion
                    col_type = CSV_HEX_TYPE_MAP[v]
```

Can we do this with dicts in Cython?

**Review by vuule — APPROVED** [review]
:fire:

File: python/cudf/cudf/_lib/csv.pyx [hunk 5]
```diff
@@ -232,37 +246,50 @@ cdef csv_reader_options make_csv_reader_options(
...
             isinstance(dtype, (
                 np.dtype, pd.core.dtypes.dtypes.ExtensionDtype, type
             ))
         ):
-            c_dtypes.reserve(1)
-            c_dtypes.push_back(
-                _get_cudf_compatible_str_from_dtype(dtype).encode()
+            c_dtypes_list.reserve(1)
+            if dtype in CSV_HEX_TYPE_MAP:
+                dtype = CSV_HEX_TYPE_MAP.get(dtype, dtype)
```
Comment (charlesbluca) [inline python/cudf/cudf/_lib/csv.pyx:268]:
```suggestion
                dtype = CSV_HEX_TYPE_MAP[dtype]
```

File: python/cudf/cudf/_lib/csv.pyx [hunk 6]
```diff
@@ -232,37 +246,50 @@ cdef csv_reader_options make_csv_reader_options(
...
+            csv_reader_options_c.set_parse_hex(c_hex_col_indexes)
         elif isinstance(dtype, abc.Iterable):
-            c_dtypes.reserve(len(dtype))
-            for col_dtype in dtype:
-                c_dtypes.push_back(
-                    _get_cudf_compatible_str_from_dtype(col_dtype).encode()
+            c_dtypes_list.reserve(len(dtype))
+            for index, col_dtype in enumerate(dtype):
+                if col_dtype in CSV_HEX_TYPE_MAP:
+                    col_dtype = CSV_HEX_TYPE_MAP.get(col_dtype, col_dtype)
```
Comment (charlesbluca) [inline python/cudf/cudf/_lib/csv.pyx:280]:
```suggestion
                    col_dtype = CSV_HEX_TYPE_MAP[col_dtype]
```

File: python/cudf/cudf/_lib/csv.pyx [hunk 7]
```diff
@@ -494,29 +521,38 @@ def _get_cudf_compatible_str_from_dtype(dtype):
...
-    elif (
-        pd_dtype in cudf.utils.dtypes.cudf_dtypes_to_pandas_dtypes or
-        str(pd_dtype) in cudf.utils.dtypes.ALL_TYPES or
-        cudf.utils.dtypes.is_categorical_dtype(pd_dtype)
-    ):
-        return str(pd_dtype)
-    else:
-        raise ValueError(f"dtype not understood: {dtype}")
+    if isinstance(dtype, str):
+        if str(dtype) == "date32":
```
Comment (charlesbluca) [inline python/cudf/cudf/_lib/csv.pyx:525]: Why the string casting here if we if `isinstance(dtype, str) == True`?

Comment (galipremsagar) [reply python/cudf/cudf/_lib/csv.pyx:525]: This is to make sure cython explicitly converts it to string type while comparison. For ex, we do this everywhere where there is an encode call or comparison call:
```python
if isinstance(col, str):
                c_parse_dates_names.push_back(str(col).encode())
```

Comment (charlesbluca) [reply python/cudf/cudf/_lib/csv.pyx:525]: Thanks for the explanation 🙂

**Review by charlesbluca — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge
